File: static/cpp/look-and-say.cc
// Copyright (c) 2013 Elements of Programming Interviews. All rights reserved.

#include <algorithm>
#include <cassert>
#include <iostream>
#include <random>
#include <sstream>
#include <string>

using std::cout;
using std::default_random_engine;
using std::endl;
using std::random_device;
using std::stoi;
using std::string;
using std::stringstream;
using std::to_string;
using std::uniform_int_distribution;

string NextNumber(const string& s);

// @include
string LookAndSay(int n) {
  string s = "1";
  for (int i = 1; i < n; ++i) {
    s = NextNumber(s);
  }
  return s;
}

string NextNumber(const string& s) {
  string result;
  for (int i = 0; i < s.size(); ++i) {
    int count = 1;
    while (i + 1 < s.size() && s[i] == s[i + 1]) {
      ++i, ++count;
    }
    result += to_string(count) + s[i];
  }
  return result;
}
// @exclude

int main(int argc, char** argv) {
  default_random_engine gen((random_device())());
  int n;
  if (argc == 2) {
    n = stoi(argv[1]);
  } else {
    uniform_int_distribution<int> dis(1, 50);
    n = dis(gen);
  }
  assert(!LookAndSay(1).compare("1"));
  assert(!LookAndSay(2).compare("11"));
  assert(!LookAndSay(3).compare("21"));
  assert(!LookAndSay(4).compare("1211"));
  assert(!LookAndSay(5).compare("111221"));
  assert(!LookAndSay(6).compare("312211"));
  assert(!LookAndSay(7).compare("13112221"));
  assert(!LookAndSay(8).compare("1113213211"));

  cout << "n = " << n << endl;
  cout << LookAndSay(n) << endl;
  return 0;
}

File: static/python/look_and_say.py
import sys
import random
import itertools


# @include
def look_and_say(n):
    def next_number(s):
        result, i = [], 0
        while i < len(s):
            count = 1
            while i + 1 < len(s) and s[i] == s[i + 1]:
                i += 1
                count += 1
            result.append(str(count) + s[i])
            i += 1
        return ''.join(result)

    s = '1'
    for _ in range(1, n):
        s = next_number(s)
    return s


# Pythonic solution uses the power of itertools.groupby().
def look_and_say_pythonic(n):
    s = '1'
    for _ in range(n - 1):
        s = ''.join(
            str(len(list(group))) + key for key, group in itertools.groupby(s))
    return s
# @exclude


def main():
    if len(sys.argv) == 2:
        n = int(sys.argv[1])
    else:
        n = random.randint(1, 50)
    assert look_and_say(1) == '1'
    assert look_and_say(2) == '11'
    assert look_and_say(3) == '21'
    assert look_and_say(5) == '111221'
    assert look_and_say(6) == '312211'
    assert look_and_say(7) == '13112221'
    assert look_and_say(8) == '1113213211'
    print('n =', n)
    print(look_and_say(n))
    assert look_and_say(n) == look_and_say_pythonic(n)


if __name__ == '__main__':
    main()
